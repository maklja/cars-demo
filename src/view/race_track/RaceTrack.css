.race-track {

	/* Race track styles */
	background-color: #eeeeee;
	margin: 0.6em 0;
	border: 0.063em solid #cccccc;
	border-radius: 1em;
	position: relative;
	min-height: 10em;
	overflow: hidden;
}

.race-track .race-track__wrapper {
	position: absolute;
	top: 0;
	left: 0;
	width: 100%;
	height: 100%;
	z-index: 2; /* put image on top track line and finish line */
}

.race-track .race-track__move {
	transition: transform 0s ease-in-out; /* easy-in-out to simulate car acceleration at the begin of animation and slow down at the end */
}

.race-track .race-track__wrapper .race-track__car {
	display: flex;
	align-items: center;
	border-radius: 1em;
	width: 12em;
	height: 100%;
}

.race-track .race-track__car.race-track__car--first-place {
	background-color: rgba(255, 223, 0, 0.8);
}

.race-track .race-track__car.race-track__car--second-place {
	background-color: rgba(192, 192, 192, 0.8);
}

.race-track .race-track__car.race-track__car--third-place {
	background-color: rgba(205, 127, 50, 0.8);
}

.race-track .race-track__wrapper .race-track__place-number {
	position: absolute;
	top: 0;
	left: 0;
	width: 100%;
	height: 100%;
	display: flex;
	justify-content: center;
	align-items: center;
	font-size: 6em;
	color: white;
}

.race-track .race-track__car .race-track__image {

	/* car image on the track */
	max-width: 100%;
	width: 100%;
}

.race-track.race-track--race-end .race-track__wrapper {
	transform: translateX(100%);
}

.race-track.race-track--race-end .race-track__wrapper .race-track__car {
	transform: translateX(-100%);
}

.race-track .race-track__line {

	/* Track lines on the middle */
	border: 0.063em dashed black;
	position: absolute;
	top: 50%;
	left: 1em;
	right: 1em;
}

.race-track .race-track__finish-line {

	/* Finish line styles made from png image */
	height: 100%;
	position: absolute;
	top: 0;
	right: 0;
	width: 2em;
	background-image: url("../../assets/images/finish_line.png");
	background-size: 200%;
}
