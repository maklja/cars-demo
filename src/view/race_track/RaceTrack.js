import PropTypes from 'prop-types';
import React, { Component } from 'react';

import { carPropTypes } from '../../props';
import { calculateTimeBySpeedAndDistance } from '../../utils';

import './RaceTrack.css';

const medalColorsClasses = [
	'race-track__car--first-place',
	'race-track__car--second-place',
	'race-track__car--third-place'
];

const RaceTrackPlace = ({ place }) => {
	return place !== -1 ? (
		<div className="race-track__place-number">
			<div>{place + 1}</div>
		</div>
	) : null;
};

RaceTrackPlace.propTypes = {
	place: PropTypes.number
};

RaceTrackPlace.defaultProps = {
	place: -1
};

export default class RaceTrack extends Component {
	constructor(props) {
		super(props);

		this._carRef = React.createRef();

		this._finishLineReached = this._finishLineReached.bind(this);
	}

	render() {
		const {
			name,
			image,
			speed,
			isRaceStarted,
			place,
			trackDistance
		} = this.props;

		const seconds = calculateTimeBySpeedAndDistance(speed, trackDistance);
		const transitionStyle = {
			// use move transition only when race starts
			// but when we do reset just immediately move
			// cars to the begin line
			transitionDuration: isRaceStarted ? `${seconds}s` : '0s'
		};
		return (
			<div
				className={`race-track ${
					isRaceStarted ? 'race-track--race-end' : ''
				}`}
				title={`Top speed: ${speed}`}
			>
				<div
					className="race-track__wrapper race-track__move"
					style={transitionStyle}
				>
					<div
						className={`race-track__car race-track__move ${this._getMedalColorsClass(
							place
						)}`}
						style={transitionStyle}
						ref={this._carRef}
					>
						<img
							className="race-track__image"
							src={image}
							alt={name}
						/>
						<RaceTrackPlace place={place} />
					</div>
				</div>
				<div className="race-track__line" />
				<div className="race-track__finish-line" />
			</div>
		);
	}

	_getMedalColorsClass(place) {
		if (place >= 0 && place < medalColorsClasses.length) {
			return medalColorsClasses[place];
		}

		return '';
	}

	componentDidMount() {
		// listen to transition end event, because thats the moment
		// when car reached finish line
		this._carRef.current.addEventListener(
			'transitionend',
			this._finishLineReached
		);
	}

	componentWillUnmount() {
		// on unmount remove listener to prevent memory leak
		// or unexpected callback triggers
		this._carRef.current.removeEventListener(
			'transitionend',
			this._finishLineReached
		);
	}

	_finishLineReached() {
		const { id, onFinishCallback } = this.props;

		onFinishCallback(id);
	}
}

RaceTrack.propTypes = {
	...carPropTypes,
	isRaceStarted: PropTypes.bool,
	place: PropTypes.number, // car place at the end of the race
	onFinishCallback: PropTypes.func, // callback will be invoked after car reaches end
	trackDistance: PropTypes.number.isRequired
};

RaceTrack.defaultProps = {
	isRaceStarted: false,
	place: -1,
	onFinishCallback: () => {}
};
