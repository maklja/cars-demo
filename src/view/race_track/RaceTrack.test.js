import React from 'react';
import renderer from 'react-test-renderer';

import RaceTrack from './RaceTrack';

describe('RaceTrack tests', () => {
	const car = {
		image:
			'https://www.mazdausa.com/siteassets/vehicles/2017/m3s/trims/sport/17-m3s-sp-25d-snowflake-white.png',
		speed: 235,
		description:
			'The Fiat Punto is a supermini car produced by the Italian manufacturer Fiat since 1993.',
		name: 'Mazda 3',
		id: 2
	};

	const createNodeMock = element => {
		// mock ref inside RaceTrack component
		if (element.props.className.indexOf('race-track__car') > -1) {
			return {
				addEventListener() {},
				removeEventListener() {}
			};
		}

		return null;
	};

	it('RaceTrack renders correctly with Mazda 3 car data', () => {
		const trackDistance = 0;

		const tree = renderer
			.create(<RaceTrack {...car} trackDistance={trackDistance} />, {
				createNodeMock
			})
			.toJSON();
		expect(tree).toMatchSnapshot();
	});

	it('RaceTrack renders correctly with gold, silver, bronze place and fourth place', () => {
		const trackDistance = 0;

		const firstPlace = 0; // gold place
		const firstPlaceRaceTrack = renderer
			.create(
				<RaceTrack
					{...car}
					place={firstPlace}
					trackDistance={trackDistance}
				/>,
				{
					createNodeMock
				}
			)
			.toJSON();
		expect(firstPlaceRaceTrack).toMatchSnapshot();

		const secondPlace = 1; // silver place
		const secondPlaceRaceTrack = renderer
			.create(
				<RaceTrack
					{...car}
					place={secondPlace}
					trackDistance={trackDistance}
				/>,
				{
					createNodeMock
				}
			)
			.toJSON();
		expect(secondPlaceRaceTrack).toMatchSnapshot();

		const thirdPlace = 2; // bronze place
		const thirdPlaceRaceTrack = renderer
			.create(
				<RaceTrack
					{...car}
					place={thirdPlace}
					trackDistance={trackDistance}
				/>,
				{
					createNodeMock
				}
			)
			.toJSON();
		expect(thirdPlaceRaceTrack).toMatchSnapshot();

		const fourthPlace = 3; // no reward place
		const fourthPlaceRaceTrack = renderer
			.create(
				<RaceTrack
					{...car}
					place={fourthPlace}
					trackDistance={trackDistance}
				/>,
				{
					createNodeMock
				}
			)
			.toJSON();
		expect(fourthPlaceRaceTrack).toMatchSnapshot();
	});
});
