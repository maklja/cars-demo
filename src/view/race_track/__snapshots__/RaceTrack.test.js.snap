// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RaceTrack tests RaceTrack renders correctly with Mazda 3 car data 1`] = `
<div
  className="race-track "
  title="Top speed: 235"
>
  <div
    className="race-track__wrapper race-track__move"
    style={
      Object {
        "transitionDuration": "0s",
      }
    }
  >
    <div
      className="race-track__car race-track__move "
      style={
        Object {
          "transitionDuration": "0s",
        }
      }
    >
      <img
        alt="Mazda 3"
        className="race-track__image"
        src="https://www.mazdausa.com/siteassets/vehicles/2017/m3s/trims/sport/17-m3s-sp-25d-snowflake-white.png"
      />
    </div>
  </div>
  <div
    className="race-track__line"
  />
  <div
    className="race-track__finish-line"
  />
</div>
`;

exports[`RaceTrack tests RaceTrack renders correctly with gold, silver, bronze place and fourth place 1`] = `
<div
  className="race-track "
  title="Top speed: 235"
>
  <div
    className="race-track__wrapper race-track__move"
    style={
      Object {
        "transitionDuration": "0s",
      }
    }
  >
    <div
      className="race-track__car race-track__move race-track__car--first-place"
      style={
        Object {
          "transitionDuration": "0s",
        }
      }
    >
      <img
        alt="Mazda 3"
        className="race-track__image"
        src="https://www.mazdausa.com/siteassets/vehicles/2017/m3s/trims/sport/17-m3s-sp-25d-snowflake-white.png"
      />
      <div
        className="race-track__place-number"
      >
        <div>
          1
        </div>
      </div>
    </div>
  </div>
  <div
    className="race-track__line"
  />
  <div
    className="race-track__finish-line"
  />
</div>
`;

exports[`RaceTrack tests RaceTrack renders correctly with gold, silver, bronze place and fourth place 2`] = `
<div
  className="race-track "
  title="Top speed: 235"
>
  <div
    className="race-track__wrapper race-track__move"
    style={
      Object {
        "transitionDuration": "0s",
      }
    }
  >
    <div
      className="race-track__car race-track__move race-track__car--second-place"
      style={
        Object {
          "transitionDuration": "0s",
        }
      }
    >
      <img
        alt="Mazda 3"
        className="race-track__image"
        src="https://www.mazdausa.com/siteassets/vehicles/2017/m3s/trims/sport/17-m3s-sp-25d-snowflake-white.png"
      />
      <div
        className="race-track__place-number"
      >
        <div>
          2
        </div>
      </div>
    </div>
  </div>
  <div
    className="race-track__line"
  />
  <div
    className="race-track__finish-line"
  />
</div>
`;

exports[`RaceTrack tests RaceTrack renders correctly with gold, silver, bronze place and fourth place 3`] = `
<div
  className="race-track "
  title="Top speed: 235"
>
  <div
    className="race-track__wrapper race-track__move"
    style={
      Object {
        "transitionDuration": "0s",
      }
    }
  >
    <div
      className="race-track__car race-track__move race-track__car--third-place"
      style={
        Object {
          "transitionDuration": "0s",
        }
      }
    >
      <img
        alt="Mazda 3"
        className="race-track__image"
        src="https://www.mazdausa.com/siteassets/vehicles/2017/m3s/trims/sport/17-m3s-sp-25d-snowflake-white.png"
      />
      <div
        className="race-track__place-number"
      >
        <div>
          3
        </div>
      </div>
    </div>
  </div>
  <div
    className="race-track__line"
  />
  <div
    className="race-track__finish-line"
  />
</div>
`;

exports[`RaceTrack tests RaceTrack renders correctly with gold, silver, bronze place and fourth place 4`] = `
<div
  className="race-track "
  title="Top speed: 235"
>
  <div
    className="race-track__wrapper race-track__move"
    style={
      Object {
        "transitionDuration": "0s",
      }
    }
  >
    <div
      className="race-track__car race-track__move "
      style={
        Object {
          "transitionDuration": "0s",
        }
      }
    >
      <img
        alt="Mazda 3"
        className="race-track__image"
        src="https://www.mazdausa.com/siteassets/vehicles/2017/m3s/trims/sport/17-m3s-sp-25d-snowflake-white.png"
      />
      <div
        className="race-track__place-number"
      >
        <div>
          4
        </div>
      </div>
    </div>
  </div>
  <div
    className="race-track__line"
  />
  <div
    className="race-track__finish-line"
  />
</div>
`;
